### PR TITLE
Restructure hypatia's ReadSet to make code more testable

### PR DESCRIPTION
**Pull request by AAmedeo**
## TLDR

Currently the ReadSets class has too many functionalities which makes it hard to set up unit tests for the hypatia code. 
These changes goal is to split the ReadSets in multiple parts that can each be tested on their own.  

## The new structure 
The main components used to replace ReadSets in the new code structure are:
1. ModelSettings -> This is used to represent the settings for a model scenario. It has 3 main functions
           -   Validate that the model settings are coherent 
           -   Store all the information historically stored in the sets folder (and offers convenient shortcuts to access commonly needed transformation of the settings)
           -   Compute based on the settings the shape of the user parameters required for the Model
2. ModelData -> This is used to validate and store the Model settings and the model user provided parameters. This is the structure storing all the Model information necessary in BuildModel to set up the cvxpy problem. 
3. Excel.py -> This is a file containing utility functions to be able to read/write settings and parameters from excel files. 

## The commits

[**Create ModelSettingData class**](https://github.com/SESAM-Polimi/Hypatia/commit/f762b84fafcb9b3fb7585a4c2374798ef70ee6f4)

*Description*
Creates ModelSetting class

*Tests*
I added unit tests for the ModelSetting. Run by:

```
python3 -m unittest hypatia/backend/tests/ModelSettingsTests.py
```

[**Add ModelData class**](https://github.com/SESAM-Polimi/Hypatia/commit/cfb656840ada2b3bccb097d6aaf17894a038bbf8)

*Description*
Creates ModelData class

*Tests*
I added unit tests for the ModelData. Run by:

```
python3 -m unittest hypatia/backend/tests/ModelDataTests.py
```

[**Add new function to interact with excel files**](https://github.com/SESAM-Polimi/Hypatia/commit/8fd29f8d96ab91212705090912793f0032e70883)

*Description*
Creates the functions to interact with excel files. Then main functions are:
1. read_settings -> Given a path to a sets folder and an execution mode, it reads settings and creates a ModelSetting structure to store them
2. write_parameters_files -> Given ModelSetting and a path, it writes the default parameters of the Model to excel files
3. read_parameters -> Given ModelSetting and a path to a parameters folder, it reads the user provided parameters and creates a ModelData object. 

*Tests*
I manually tested these functions and made sure that I was able to correctly read and write excel files. 
For now I have not written automated tests for these function, but we can consider it in the future if it seems necessary. 

[**Switch model building to use ModelData instead of ReadSets**](https://github.com/SESAM-Polimi/Hypatia/commit/5a5487f297a096634cc63316dd17fccec7023a5a)

*Description*

This switches all the code to use ModelData instead of ReadSets. 

*Tests*
For now I limited myself to manually run hypatia (https://gist.github.com/AAmedeo/637b33c80b44935be7fe57f846bd5fa1) on the existing example models and made sure that the result stayed the same. 
However, we are planning to add lots of new tests cases in the near future to verify the correctness of the optimization results returned by hypatia. 

[**Move create_technology_columns to utilities**](https://github.com/SESAM-Polimi/Hypatia/commit/6abae06acdfc502e3edfcd34077a9c556d50dd2e)

*Description*
Moved create_technology_columns function out of StrData.py since I am planning to remove that file in the following diff 

*Tests*
I just run the existing unit tests
```
python3 -m unittest hypatia/utility/tests/utilityTest.py
python3 -m unittest hypatia/backend/tests/ModelSettingsTests.py
```

[**Fix plotting to use ModelData**](https://github.com/SESAM-Polimi/Hypatia/commit/879138f49af543b16635d55ec89863bdbd1e30ac)

*Description*
Make sure that the plotting library uses ModelData instead of ReadSets

*Tests* 
I tried to create various plot and verified that the plot worked correctly. 
For example running a script like https://gist.github.com/AAmedeo/0bf7814e8e00ff9cacda0aaef3825aee

For now I have not written automated tests for the plotting function, but I think we should look into restructuring and testing
the plotting functionality in the near future.  

[**Remove StrData**](https://github.com/SESAM-Polimi/Hypatia/commit/594e0ff01c45810028a0fece4d8b4c04259e203a)
*Description*
Remove unused code

*Tests*
The code is already unused so it shouldn't effect anything else. 
I run all existing automated tests:
```
python3 -m unittest hypatia/utility/tests/utilityTest.py
python3 -m unittest hypatia/backend/tests/ModelSettingsTests.py
python3 -m unittest hypatia/backend/tests/ModelDataTests.py
```
Also I run did some manual tests to verify that the new restructured code behaves the same as before. 
I did this by running this [<script> ](https://gist.github.com/AAmedeo/637b33c80b44935be7fe57f846bd5fa1)before and after the changes in this pull request and made sure the output is the same.

### CODE REVIEW

**Review by Negnamazi — CHANGES_REQUESTED** [review]
@AAmedeo Nice to see that the function for reading/writing the excel files are separated. Thanks ! Can we just avoid using the __get_number_of_levels() as it might be a bit risky because sometime index.names can be "None" as it is an optional property of pd.Index and pd.MultiIndex objects. There is another property of these two Classes that returns what we need: pd.Index.nlevels, pd.MultiIndex.nlevels

**Comment by AAmedeo** [comment]
Thank you for the review :)

I updated the code to use nlevels instead of __get_number_of_levels() based on your suggestion. 
In order to make it work I had to fix some of the parameters definitions since they were setting the "columns" field  to List/Series which does not have nlevels. After this change all the "columns" and "index" fields should only be set to pd.Index or pd.MultiIndex. 

I tested the change manually running: https://gist.github.com/AAmedeo/6ec78988cb206cf28a92dd0ad1b5f386
